File: internal/packet/ca_login.go
package packet

import (
	"fmt"
	"io"

	"github.com/macabu/cronus/pkg/xbinary"
)

type PACKET_CA_LOGIN struct {
	// /* this+0x0 */ short PacketType
	version    uint32 // /* this+0x2 */ unsigned long Version
	id         string // /* this+0x6 */ unsigned char ID[24]
	password   string // /* this+0x1e */ unsigned char Passwd[24]
	clienttype uint8  // /* this+0x36 */ unsigned char clienttype
}

func (p *PACKET_CA_LOGIN) Size() int {
	return 55
}

func (p *PACKET_CA_LOGIN) ReadFrom(r io.Reader) (int64, error) {
	if err := xbinary.Read(r, &p.version); err != nil {
		return 0, fmt.Errorf("reading version: %w", err)
	}

	if err := xbinary.ReadString(r, &p.id, 24); err != nil {
		return 0, fmt.Errorf("reading id: %w", err)
	}

	if err := xbinary.ReadString(r, &p.password, 24); err != nil {
		return 0, fmt.Errorf("reading password: %w", err)
	}

	if err := xbinary.Read(r, &p.clienttype); err != nil {
		return 0, fmt.Errorf("reading clienttype: %w", err)
	}

	return int64(p.Size()), nil
}

File: src/packet/ca_login.rs
#![allow(non_camel_case_types)]
use super::PacketHeader;

/*
// packet 0x64
struct PACKET_CA_LOGIN {
  /* this+0x0 */ short PacketType
  /* this+0x2 */ unsigned long Version
  /* this+0x6 */ unsigned char ID[24]
  /* this+0x1e */ unsigned char Passwd[24]
  /* this+0x36 */ unsigned char clienttype
}
*/
#[derive(Debug, Clone)]
pub struct PACKET_CA_LOGIN {
    pub version: u32,
    pub id: String,
    pub password: String,
    pub clienttype: u8,
}

impl PacketHeader for PACKET_CA_LOGIN {
    const HEADER: i16 = 0x64;
}

impl TryFrom<&[u8]> for PACKET_CA_LOGIN {
    type Error = Box<dyn std::error::Error>;

    fn try_from(buf: &[u8]) -> Result<Self, Self::Error> {
        let mut offset = 2;

        let version = u32::from_le_bytes(buf[offset..4 + offset].try_into()?);
        offset += 4;

        let id = String::from_utf8_lossy(&buf[offset..24 + offset]).into_owned();
        offset += 24;

        let password = String::from_utf8_lossy(&buf[offset..24 + offset]).into_owned();
        offset += 24;

        let clienttype = buf[offset..1 + offset][0];

        Ok(PACKET_CA_LOGIN {
            version,
            id,
            password,
            clienttype,
        })
    }
}
